virtio-devices: seccomp: Add seccomp_filter module

likebreath opened the pull request:
This patch added the seccomp_filter module to the virtio-devices crate
by taking reference code from the vmm crate. This patch also adds
allowed-list for the virtio-block worker thread.

Partially fixes: #925

This patch also extends the CLI option '--seccomp' to accept the 'log'
parameter in addition 'true/false'. It also refactors the
vmm::seccomp_filters module to support both "SeccompAction::Trap" and
"SeccompAction::Log".

Fixes: #1180

Signed-off-by: Bo Chen <chen.bo@intel.com>

likebreath commented:
I have reproduced the error locally, and will update the PR accordingly tomorrow.

likebreath commented:
@sboeuf @rbradford Any advises on how to reproduce the errors on "aarch64 worker build" (which is the only failing CI job now).

sboeuf commented:
@likebreath looking at the failures, I'd say this PR broke any `aarch64` test that relies on a `virtio-blk` device, which is unfortunate :)
I think you're simply missing some syscalls in the list of authorized syscalls. Problem is, I don't have an `aarch64` machine to reproduce and help identify what's missing :(
@michael2012z @MrXinWang could you help us with this task?

MrXinWang commented:
@sboeuf Can you please re-trigger the CI again so we can confirm if this CI failure was due to our yesterday's debug of https://github.com/cloud-hypervisor/cloud-hypervisor/issues/1505 and https://github.com/cloud-hypervisor/cloud-hypervisor/issues/1507? Sorry for the inconvenience...

sboeuf commented:
@MrXinWang 
>@sboeuf Can you please re-trigger the CI again so we can confirm if this CI failure was due to our yesterday's debug of #1505 and #1507? Sorry for the inconvenience...

Oh of course! No worries :)

MrXinWang commented:
Thanks @sboeuf. One thing I need to note is that to fix https://github.com/cloud-hypervisor/cloud-hypervisor/issues/1516, I just changed the host kernel of the AArch64 CI machine. Donno this will affect the jenkins...

> Problem is, I don't have an aarch64 machine to reproduce and help identify what's missing :(
Can you please also check your slack please :)

sboeuf commented:
@MrXinWang just curious, why did you need to update the host kernel to >5.5 ?

MrXinWang commented:
@sboeuf As there is a AArch64 bug and this bug was fixed by one of my colleague and then merged in kernel 5.5. To enable virtio-fs, we need this bug fixed...

MrXinWang commented:
> One thing I need to note is that to fix #1516, I just changed the host kernel of the AArch64 CI machine. Donno this will affect the jenkins... I can still ssh to it now though.

Hi @rbradford, sorry to interrupt but do you know how can I wake the Jenkins agent up for this machine? Thanks!

sboeuf commented:
@likebreath @MrXinWang testing on the `aarch64` machine showed me that `virtio-blk` is working as expected with the current set of seccomp filters defined from this PR.
So I guess this was just a glitch from the `aarch64` CI. Let's cross our fingers the CI will turn green :)

rbradford commented:
@MrXinWang The SSH host keys on the machine had changed so it needed refreshing

MrXinWang commented:
@rbradford I re-generated the ssh key for the `jenkins` account.

The new public key is:
```
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDAHPqdS3JhYZxTDr69DHA59afrqin7G9Vi9RlzaCggYGvkhxjxj3T4WGQo+UpNlP+7U1cIZIP2sCkfDtx83fL7YuCYvmc0csMt3icephDwWJWpQqbnQFckoVvzK9RHf6BP8+cWcQ5LP82Vfrmi4sv7KC8HVPsAz3nRGAPiyaaQ7Qlxo9AlmPQNl8caTVTYIT5AGruqJLGKpZ57Fb5dW5W37hpbLbyco+7CT1LGRPDnGEHqGd/XQrEBqW9kR5t/NF1ZMPCB2XGM63BmDKT1zNqzY1t1IMMR+lcHhRSVWY6w1gE9qgeGbw7gDjxnXRHTRWtcLojGVR10VdJYR/k9roFL jenkins@ci-cloud-hypervisor-arm64-01
```

authorized_keys:
```
ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBCOZZUgHnUYHZ/U0rhPUW8tdslqwvbj3LEzbeJg5Gg2HsdizpIMamlb6wqb7Tvz1IQz7J3YOOXKNzuKhhk25whM= jenkins@jenkins
```

rbradford commented:
> @rbradford I re-generated the ssh key for the `jenkins` account.
> 
> The new public key is:
> 
> ```
> ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDAHPqdS3JhYZxTDr69DHA59afrqin7G9Vi9RlzaCggYGvkhxjxj3T4WGQo+UpNlP+7U1cIZIP2sCkfDtx83fL7YuCYvmc0csMt3icephDwWJWpQqbnQFckoVvzK9RHf6BP8+cWcQ5LP82Vfrmi4sv7KC8HVPsAz3nRGAPiyaaQ7Qlxo9AlmPQNl8caTVTYIT5AGruqJLGKpZ57Fb5dW5W37hpbLbyco+7CT1LGRPDnGEHqGd/XQrEBqW9kR5t/NF1ZMPCB2XGM63BmDKT1zNqzY1t1IMMR+lcHhRSVWY6w1gE9qgeGbw7gDjxnXRHTRWtcLojGVR10VdJYR/k9roFL jenkins@ci-cloud-hypervisor-arm64-01
> ```
> 
> authorized_keys:
> 
> ```
> ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBCOZZUgHnUYHZ/U0rhPUW8tdslqwvbj3LEzbeJg5Gg2HsdizpIMamlb6wqb7Tvz1IQz7J3YOOXKNzuKhhk25whM= jenkins@jenkins
> ```
> 
> Are these two keys enough for refreshing the agent?

No-no. It was the host key that was out of date and I had already fixed it when I replied.

likebreath commented:
As advised by @rbradford, I made two changes to the PR:
1. Added a new allowed syscall for AArch64:  `allow_syscall(46)` as alternative for `ftruncate`.
2. Split the PR into two commits.

Let's see whether the CI is happy, if not I will need to debug it on the AArch64 machine.

likebreath commented:
@rbradford @sboeuf PTAL. CI is passing now.

rbradford commented on vmm/src/api/openapi/cloud-hypervisor.yaml:
```diff
@@ -417,6 +417,9 @@ components:
           default: false
+        seccomp:
+          type: boolean
+          default: true
```
I don't think this should be here. The secomp control is part of the VMM control rather than for the VM.

sboeuf replied:
I agree, but I can also see why @likebreath took this approach. We don't really have a way to tell the VMM if we want seccomp to be enabled/disabled through the OpenAPI.

rbradford commented on src/main.rs:
```diff
@@ -261,7 +261,8 @@ fn create_app<'a, 'b>(
                 .possible_values(&["true", "false"])
-                .default_value("true"),
+                .default_value("true")
+                .group("vm-config"),
```
seccomp filters are dealt with at the VMM level so although we need to propagate the value into the vm and device manager it shouldn't be configurable per-vm.

sboeuf replied:
Agreed.

likebreath commented:
@rbradford @sboeuf PR updated for addressing the comments above. PTAL. Thanks!

sboeuf commented on src/main.rs:
```diff
@@ -288,18 +287,17 @@ fn start_vmm(cmd_arguments: ArgMatches) {
-            "true" => SeccompLevel::Advanced,
-            "false" => SeccompLevel::None,
+            "true" => true,
+            "false" => false,
```
No we don't want to switch to a boolean model, as we'd like to introduce a third option to log the faulty syscalls.

likebreath commented:
@sboeuf @rbradford I updated the PR which also addressed issue #1180 (allow to log faulty syscalls). PTAL.
